### PR TITLE
New CSS selector  ":contains(text)" 

### PR DESCRIPTION
**Pull request by f34nk**
Hey Alex,
I followed your instructions in [#36](https://github.com/lexborisov/Modest/issues/36) and added new pseudo class "contains".

- added "contains" to `$func_map = {...}`
- run script `perl utils/MyCSS/functions.pl` 
- added all "contains" related functions and enums
- function `mycss_selectors_function_begin_contains()` in file **function.c**:
added selector sub_type `MyCSS_SELECTORS_SUB_TYPE_PSEUDO_CLASS_FUNCTION_CONTAINS`
- added test `test/myhtml/pseudo_class_contains.c`
- implemented finder function in `modest_finder_selector_sub_type_pseudo_class_function_contains()` in file **pseudo_class.c**.

As of now, the selector argument cannot contain single or double quotation marks. That's because I basically duplicated the implementation from pseudo class "has", so behavior and serialization is the same.

Example:
```
const char *selector = ":contains(Hello)"; // works
const char *selector = ":contains('Hello')"; // Bad Selector
const char *selector = ":contains(\"Hello\")"; // Bad Selector
```

According to the [spec](https://www.w3.org/TR/2001/CR-css3-selectors-20011113/#content-selectors) (from 2001) however, quotation marks should also be supported.
> The :contains("foo") pseudo-class notation represents an element whose textual contents contain the given substring. The argument of this pseudo-class can be a string (surrounded by double quotes) or a keyword.

So my implementation only solves the "argument as keyword" case. 
Tell me what you think.

Best, 
Frank

### CODE REVIEW

File: source/modest/finder/finder.c [hunk 1]
```diff
@@ -189,9 +190,9 @@ myhtml_tree_node_t * modest_finder_node_combinator_begin(modest_finder_t* finder
                 if(callback_found)
                     callback_found(finder, node, selector_list, selector, spec, ctx);
             }
-            else {
+            else {
```
Comment (lexborisov) [inline source/modest/finder/finder.c]: Extra spaces after `else {`

File: source/modest/finder/finder.c [hunk 2]
```diff
@@ -229,6 +230,10 @@ myhtml_tree_node_t * modest_finder_node_combinator_undef(modest_finder_t* finder
     if(selector == NULL)
         return NULL;
     
+    // FRANK
```
Comment (lexborisov) [inline source/modest/finder/finder.c]: remove this

File: source/modest/finder/finder.c [hunk 3]
```diff
@@ -229,6 +230,10 @@ myhtml_tree_node_t * modest_finder_node_combinator_undef(modest_finder_t* finder
...
+    // printf("\nmodest_finder_node_combinator_undef()\n");
+    // printf("\tselector->type = %d\n", (int)selector->type);
+
```
Comment (lexborisov) [inline source/modest/finder/finder.c]: need to remove comments above

File: source/modest/finder/pseudo_class.c [hunk 4]
```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
     return false;
 }
 
+// concat str1 and str2
+char *concat_string(const char *str1, const char *str2)
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: move to `core` module, append to [mycore/utils.c](https://github.com/lexborisov/Modest/blob/master/source/mycore/utils.c) and rename to `mycore_str_cat`

File: source/modest/finder/pseudo_class.c [hunk 5]
```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+        if(str1) strcpy(result, str1);
+        if(str2) strcat(result, str2);
+    }
+
+    return result;
+}
+
+bool modest_finder_selector_sub_type_pseudo_class_function_contains(modest_finder_t* finder, myhtml_tree_node_t* base_node, mycss_selectors_entry_t* selector, mycss_selectors_specificity_t* spec)
+{
+    if(base_node){
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]:
```C
if(base_node == NULL) {
    return false;
}
```

File: source/modest/finder/pseudo_class.c [hunk 6]
```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+// concat str1 and str2
+char *concat_string(const char *str1, const char *str2)
+{
+    char *result = NULL;
+    size_t n = 0;
+
+    if(str1) n += strlen(str1);
+    if(str2) n += strlen(str2);
+
+    if((str1 || str2) && (result = malloc(n + 1)) != NULL)
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: Too many checks of the same `str1` and `str2`
You can make it easier

Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]:
```C
result = malloc(n + 1);
if(result == NULL) {
    return NULL;
}
```

File: source/modest/finder/pseudo_class.c [hunk 7]
```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+                mycss_selectors_entry_t *sel_entry = list->entries_list[i].entry;
+
+                if(sel_entry->key->data){
+                    data = concat_string(data, sel_entry->key->data);
+                }
+                
+                mycss_selectors_entry_t *next = sel_entry->next;
+                while(next && !i_found){
+                    if(next->key->data){
+                        data = concat_string(data, " ");
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: memory leak!
if we allocation memory in `concat_string` and call `concat_string` with returned data we get leak.

File: source/modest/finder/pseudo_class.c [hunk 8]
```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+                        data = concat_string(data, next->key->data);
+                    }
+                    next = next->next;
+                }
+            }
+            if(strstr(text, data) != NULL) {
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: Where is the release (`mycore_free(data)`) of resources for `data`?

File: source/mycss/selectors/function.c [hunk 9]
```diff
@@ -34,7 +34,7 @@ const mycss_selectots_function_begin_entry_t * mycss_function_begin_entry_by_nam
                   % MyCSS_SELECTORS_FUNCTION_NAME_STATIC_SIZE) + 1;
     
     while (mycss_selectors_function_begin_map_index[idx].name)
-    {
+    {
```
Comment (lexborisov) [inline source/mycss/selectors/function.c]: unnecessary spaces after `{`

File: source/mycss/selectors/function.c [hunk 10]
```diff
@@ -110,6 +110,25 @@ void mycss_selectors_function_begin_has(mycss_entry_t* entry, mycss_selectors_en
...
+    
+    selectors->list = new_list;
+    selectors->list_last = NULL;
+    selectors->ending_token = entry->parser_ending_token;
+    
+    mycss_selectors_state_relative_selector_list(entry, NULL, true);
+    
+    (*new_list)->parent = current_list;
+    
+    mycss_entry_parser_list_push(entry, mycss_selectors_function_parser_has, entry->parser_switch, entry->parser_ending_token, false);
```
Comment (lexborisov) [inline source/mycss/selectors/function.c]: Why didn't you implement their parsing function?
Do not use `mycss_selectors_function_parser_has` in this.

File: source/modest/finder/finder.c (hunk 1 above)
Comment (f34nk) [reply source/modest/finder/finder.c]: ok

File: source/modest/finder/finder.c (hunk 3 above)
Comment (f34nk) [reply source/modest/finder/finder.c]: ok

File: source/modest/finder/pseudo_class.c (hunk 4 above)
Comment (f34nk) [reply source/modest/finder/pseudo_class.c]: ok

File: source/modest/finder/pseudo_class.c (hunk 5 above)
Comment (f34nk) [reply source/modest/finder/pseudo_class.c]: ok

File: source/modest/finder/pseudo_class.c (hunk 7 above)
Comment (f34nk) [reply source/modest/finder/pseudo_class.c]: Of course :)

File: source/mycss/selectors/function.c (hunk 9 above)
Comment (f34nk) [reply source/mycss/selectors/function.c]: ok

File: source/mycss/selectors/function.c (hunk 10 above)
Comment (f34nk) [reply source/mycss/selectors/function.c]: ok

File: source/modest/finder/pseudo_class.c [hunk 11]
```diff
@@ -77,6 +77,57 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+                snprintf(&data[prev], length, "%s%s", whitespace, str);
+            }
+            next = next->next;
+        }
+
+        if(strstr(text, data) != NULL) {
+            mycore_free(data);
+            return true;
+        }
+        else {
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: unnecessary `else`

```C
if(strstr(text, data) != NULL) {
    mycore_free(data);
    return true;
}

mycore_free(data);

return false;
```

File: source/modest/finder/pseudo_class.c [hunk 12]
```diff
@@ -77,6 +77,57 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+        return false;
+
+    const char* text = myhtml_node_text(text_node, NULL);
+    if(text == NULL)
+        return false;
+
+    mycss_selectors_list_t *list = (mycss_selectors_list_t*)selector->value;
+    for(size_t i = 0; i < list->entries_list_length; i++) {
+        char *data = NULL;
+        data = mycore_malloc(0);
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]:
```C
if(data == NULL) {
    return false;
}
```

File: source/modest/finder/pseudo_class.c [hunk 13]
```diff
@@ -77,6 +77,57 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+
+        mycss_selectors_entry_t *sel_entry = list->entries_list[i].entry;
+        if(sel_entry->key->data){
+            const char *str = sel_entry->key->data;
+            int length = strlen(str) + 1;
+            data = mycore_realloc(data, length);
+            snprintf(&data[0], length, "%s", str);
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: leak detected if `mycore_realloc` return `NULL`. Need check.
```C
if(sel_entry->key->data){
    const char *str = sel_entry->key->data;
    int length = strlen(str) + 1;

    char *new_data = mycore_realloc(data, length);
    if(new_data == NULL) {
        mycore_free(data);
        return false;
    }

    snprintf(&data[0], length, "%s", str);
}
```

File: source/modest/finder/pseudo_class.c [hunk 14]
```diff
@@ -77,6 +77,57 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+
+        mycss_selectors_entry_t *next = sel_entry->next;
+        while(next) {
+            if(next->key->data) {
+                int prev = strlen(data);
+                const char* whitespace = (prev > 0) ? " " : "";
+                const char *str = next->key->data;
+                int length = strlen(whitespace) + strlen(str) + 1;
+                data = mycore_realloc(data, prev + length);
+                snprintf(&data[prev], length, "%s%s", whitespace, str);
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: and this we have memory leak. see prev comment

File: source/modest/finder/pseudo_class.c [hunk 15]
```diff
@@ -77,6 +77,65 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+        data = mycore_malloc(0);
+        if(data == NULL) {
+            return false;
+        }
+        mycss_selectors_entry_t *sel_entry = list->entries_list[i].entry;
+        if(sel_entry->key->data){
+            const char *str = sel_entry->key->data;
+            int length = strlen(str) + 1;
+            data = mycore_realloc(data, length);
+            if(data == NULL) {
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: leak detected, see [comment](https://github.com/lexborisov/Modest/pull/42#discussion_r177076853)

File: source/modest/finder/pseudo_class.c [hunk 16]
```diff
@@ -77,6 +77,68 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+        char *data = NULL;
+        data = mycore_malloc(0);
+        if(data == NULL) {
+            return false;
+        }
+        
+        mycss_selectors_entry_t *sel_entry = list->entries_list[i].entry;
+        if(sel_entry->key->data){
+            const char *str = sel_entry->key->data;
+            int length = strlen(str) + 1;
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: add new line after this

File: source/modest/finder/pseudo_class.c [hunk 17]
```diff
@@ -77,6 +77,68 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+            char *new_data = mycore_realloc(data, length);
+            if(new_data == NULL) {
+                mycore_free(data);
+                return false;
+            }
+            snprintf(&new_data[0], length, "%s", str);
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: new line before

Comment (lexborisov) [reply source/modest/finder/pseudo_class.c]: and change to `snprintf(new_data, length, "%s", str);`

File: source/modest/finder/pseudo_class.c [hunk 18]
```diff
@@ -77,6 +77,68 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+            data = new_data;
+        }
+
+        mycss_selectors_entry_t *next = sel_entry->next;
+        while(next) {
+            if(next->key->data) {
+                int prev = strlen(data);
+                const char *whitespace = (prev > 0) ? " " : "";
+                const char *str = next->key->data;
+                int length = strlen(whitespace) + strlen(str) + 1;
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: new line after

File: source/modest/finder/pseudo_class.c [hunk 19]
```diff
@@ -77,6 +77,68 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
...
+                char *new_data = mycore_realloc(data, prev + length);
+                if(new_data == NULL) {
+                    mycore_free(data);
+                    return false;
+                }
+                snprintf(&new_data[prev], length, "%s%s", whitespace, str);
```
Comment (lexborisov) [inline source/modest/finder/pseudo_class.c]: new line before

**Comment by lexborisov** [comment]
AWESOME!

**Comment by f34nk** [comment]
Thanks man! That was fun :)

**Pull request merged**